src/Intervals.php

```diff
@@ -0,0 +1,371 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\EmptyConstraint;
+use Composer\Semver\Constraint\MultiConstraint;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class Intervals
+{
+    private static $intervalsCache = array();
+
+    /**
+     * @psalm-var array<string, int>
+     */
+    private static $opSortOrder = array(
+        '>=' => -3,
+        '<' => -2,
+        '==' => -1,
+        '!=' => 1,
+        '>' => 2,
+        '<=' => 3,
+    );
+
+    public static function clear()
+    {
+        self::$intervalsCache = array();
+    }
+
+    public static function isSubsetOf(ConstraintInterface $candidate, ConstraintInterface $constraint)
```

Comment:
you should add phpdoc on all public methods to document their return type (as we cannot put it in the signature itself due to supporting PHP 5)